mgr/dashboard: clean-up python unit tests

epuertat opened the pull request:
## Dashboard Python Unit Test Refactoring

Now Python unit tests don't require Ceph to built before in order to run. All dependencies have been removed or mocked.

Unit-testing:
- Add conftest.py to mock imported modules (rados, rbd, cephfs), and mock
also rados Error and OSError Exceptions.
- Fix pylint issues.
- Fix RST issues (HACKING.rst)
- Refactor JwtManager check to avoid code duplication (found by pylint when --jobs=1).
- Fix issue with DashboardException.code property, using abs() on potentially None attribute.
- Disables Doctests in services/rbd.py that are actually integration tests (they check the value of rbd.RBD_FEATURES_NAME_MAPPING). Ideally, these kind of tests should be explicitly executed in an integration testing stage, rather that unit-testing. Disabled tests have been prepended with @DISABLEDOCTEST token.

Refactor tox.ini:
- Remove requirements-{py27,py3}.txt, as python release dependant
packages can be handled with PEP 508 syntax.
- Remove develepment dependencies from requirements.
- Move pycodestyle settings to separate section.
- Add flake8 check
- Add RST linter.
- Add instafail plugin to report failures immediately
- Add timeout plugin to limit max run time (sometimes test_tasks hangs)
- Remove unused dependencies (lru_cache, pluggy)

Docs:
 - Update HACKING.rst

Fixes: https://tracker.ceph.com/issues/40487
Fixes: https://tracker.ceph.com/issues/41152
Signed-off-by: Ernesto Puerta <epuertat@redhat.com>

---

- [x] References tracker ticket
- [x] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug

alfonsomthd commented on src/pybind/mgr/dashboard/services/rbd.py:
```diff
@@ -25,7 +25,7 @@ def format_bitmask(features):
     Formats the bitmask:
 
-    >>> format_bitmask(45)
+    @DISABLEDOCTEST: >>> format_bitmask(45)
```
Has this been disabled on purpose?

epuertat replied:
Yup, these doctests require `rados` and `rbd` packages, so in my opinion they're integration tests (they are testing outputs from other modules rather than from the function itself). Ideally some mocking should be done here, but the truth is that mocking in doctests is a really cumbersome thing.

alfonsomthd replied:
OK, then should it make sense to move them to the appropriate test files or even are they duplicated?

alfonsomthd commented on src/pybind/mgr/dashboard/tox.ini:
```diff
@@ -1,29 +1,60 @@
-    lint: pycodestyle --max-line-length=100 --exclude=.tox,venv,frontend,.vscode --ignore=E402,E121,E123,E126,E226,E24,E704,W503,E741 .
-    run: {posargs}
+commands =
+    cov: pytest --cov --cov-append --cov-report=term-missing --doctest-modules --ignore=module.py --ignore=frontend/ {posargs}
```
Can we get rid of the "Missing" column in coverage report?
IMHO is a little bit overwhelming:
```
  Name                                   Stmts   Miss  Cover   Missing
--------------------------------------------------------------------
awsauth.py                                80     57    29%   75-84, 87-96, 99-183
conftest.py                               12      0   100%
controllers/__init__.py                  545     79    86%   31, 34, 37, 40, 54, 77, 80, 137-139, 151, 154, 197, 204, 220, 251-252, 266, 280-282, 306, 308, 336-354, 406, 435-438, 449, 483-486, 564, 575, 583-596, 653-654, 663, 667, 671, 675-677, 681-684, 688, 773, 785, 796, 811, 814, 867, 891-894, 899-900
controllers/auth.py                       37     21    43%   19, 21-33, 40-45, 50, 52, 57-63
controllers/cephfs.py                    140    117    16%   19-23, 26-27, 30-32, 36-38, 50-77, 81-84, 89-101, 105-234, 247-279, 287-288, 292
controllers/cluster_configuration.py      66     45    32%   22-32, 35-36, 39, 43-55, 59-66, 69-72, 77, 80-83, 87-91, 94-102
controllers/docs.py                      217     50    77%   55, 69, 71, 73, 75, 90-97, 124-135, 157-160, 162, 312-329, 369-390, 451
controllers/erasure_code_profile.py       38      8    79%   41-42, 45-46, 50, 56-58
controllers/grafana.py                    27      2    93%   29, 43
controllers/health.py                    121     73    40%   25, 29, 36-64, 68-79, 83-90, 95-103, 106, 109-118, 121, 124, 129, 131-132, 136, 138-143, 147-161, 167-169, 173, 182, 188
controllers/host.py                        8      1    88%   12
controllers/iscsi.py                     474     32    93%   33-43, 46, 51-59, 62, 67, 73, 91, 112, 119-120, 136, 139, 348, 652
controllers/logging.py                     7      1    86%   10
controllers/logs.py                       32     17    47%   18-21, 24-27, 30-33, 36-42, 47-48
controllers/mgr_modules.py                77     55    29%   22-36, 46-52, 63-67, 77-78, 89-90, 102-103, 114-120, 131-132, 143-144, 150-168
controllers/monitor.py                    24     15    38%   16-36
controllers/nfsganesha.py                123     70    43%   98-105, 116-121, 131-153, 162-166, 177-209, 223-231, 248-266, 273, 277, 281-297, 301, 305, 309
controllers/osd.py                       107     43    60%   23, 29, 35, 43, 57-64, 74-78, 92, 101, 105, 122, 133, 145, 158, 170, 178-182, 187-188, 199-206, 209, 217-229
controllers/perf_counters.py              51     18    65%   16-38, 85
controllers/pool.py                      115     83    28%   24-44, 47-55, 58, 62-66, 70-72, 77, 82-83, 89-94, 97-131, 134-141, 146, 154-182
controllers/prometheus.py                 34      0   100%
controllers/rbd.py                       307    169    45%   37, 42-45, 59-61, 64, 81-86, 89, 92, 98-101, 105-116, 121, 126-127, 130, 137-151, 158-182, 187-198, 201-213, 218, 223-227, 234, 240-250, 256-257, 261-265, 275-280, 285-294, 307-310, 314-326, 333-336, 340-341, 350, 356, 367-370, 379, 387-389, 395-397, 404-406, 423-427, 437-442, 452-455, 463, 466-473, 479, 488-492, 501, 509
controllers/rbd_mirroring.py             254     85    67%   54, 124, 128, 132, 153-167, 171, 174, 251, 254-255, 279, 288-307, 323-324, 351, 353-356, 360, 364-372, 382, 384,
[...]
```

epuertat replied:
Removed. I liked it because it allows you to easily work on improving coverage without the need of using the coverage.xml/html reports.

epuertat commented:
jenkins retest this please

epuertat commented:
jenkins retest this please

epuertat commented:
jenkins retest this please

sebastian-philipp commented:
> Disables Doctests in services/rbd.py that are actually integration tests
(they check the value of rbd.RBD_FEATURES_NAME_MAPPING). Ideally, these
kind of tests should be explicitly executed in an integration testing
stage, rather that unit-testing. Disabled tests have been prepended with
@DISABLEDOCTEST token.

Actually I have to disagree here. Integration tests are tests that depend on a running cluster. Just importing another python module from the ceph tree doesn't make this an integration tests.

Well, I can see the desire to no longer depend on compiled ceph libs. What about introducing a command line flag to optionally enable those tests in `make check`?

epuertat commented:
> Well, I can see the desire to no longer depend on compiled ceph libs. What about introducing a command line flag to optionally enable those tests in `make check`?

Mmm, I'm afraid that is a misconception coming from the fact that we have everything versioned into a single repo, which does not enforce a clear separation of concerns, and also a "make" that makes _literally everything_ (including launching a local cluster).

Integration testing does not require any cluster to run, as long as you find a better/faster way to test the interfaces between components, which is the sole purpose of integration testing.

In this case, those doctests that I disabled barely checked the internal behaviour of the functions, but mostly the values returned by external dependencies (rbd Python bindings in this case). I'm not against integration tests, just the opposite, but let's keep unit tests small, independent, quick to build and run.

sebastian-philipp commented:
> let's keep unit tests small, independent, quick to build and run.

:100: 

> > Well, I can see the desire to no longer depend on compiled ceph libs. What about introducing a command line flag to optionally enable those tests in `make check`?
> 
> Mmm, I'm afraid that is a misconception coming from the fact that we have everything versioned into a single repo, which does not enforce a clear separation of concerns, and also a "make" that makes _literally everything_ (including launching a local cluster).

Of course, everything else would be wrong. `make` is supposed to make everything.

> 
> Integration testing does not require any cluster to run, as long as you find a better/faster way to test the interfaces between components, which is the sole purpose of integration testing.

So, running this doc test is not _slow_. Actually it's super fast, if you have compiled ceph libs.


> In this case, those doctests that I disabled barely checked the internal behaviour of the functions, but mostly the values returned by external dependencies (rbd Python bindings in this case).

Hm. To me this sounds like a unit test.

epuertat commented:
> > Mmm, I'm afraid that is a misconception coming from the fact that we have everything versioned into a single repo, which does not enforce a clear separation of concerns, and also a "make" that makes _literally everything_ (including launching a local cluster).
> 
> Of course, everything else would be wrong. `make` is supposed to make everything.

That's my very point against mono-repos, as they allow/enforce this way of thinking towards code: a big black chaotic box, where a line of code here can/may/might _break_ something else over there. `make` is not and shouldn't be supposed to test everything: `make` is supposed to build code (which is in itself a primary test), ensure a basic degree of sanity testing (unit, component, some integration testing perhaps), and that's all.

And, even worse, `make` doesn't test everything. It doesn't recompile the kernel, or the firmware, or the very compiler, and changes happen also there. So, in the end, we are still using understanding and trust to narrow the scope of uncertainty. That's the whole point of "modern" (in the 90s) software practices: modularity, interfaces, dependencies, and delimiting the impact of changes so you just rebuild and test what it changed.

> So, running this doc test is not _slow_. Actually it's super fast, if you have compiled ceph libs.

The trick lies in the "if you have compiled..." clause, which by default equals to `False` (and according to the "build-everything-on-every-change" philosophy, it should be `False` always). So the difference by commenting these 4 doctests (which are trivial and barely test anything in dashboard code), is ~20 minutes build vs. no build at all.

I wouldn't comment out these doctests if we took librbd python bindings from Pypi (no trace of it though) and use it directly (though I still think they fall under the definition of integration tests).

ricardoasmarques commented on src/pybind/mgr/dashboard/module.py:
```diff
@@ -489,7 +489,7 @@ def index(self):
                 if active_uri:
                     module.log.info("Redirecting to active '%s'", active_uri)
-                    raise cherrypy.HTTPRedirect(active_uri)
+                    raise cherrypy.HTTPError(500, message="Unavailable")
```
This should be reverted

epuertat replied:
Thanks! Removed.

epuertat commented:
As per discussed in biweekly grooming meeting:
- Split into 2 different commits:
  - Fixes/refactoring to the unit tests
  - Refactoring to tox and related stuff
- Removed CircleCI support
- `install-deps.sh` python package pre-caching is kept removed. I'll add a comment below for further discuss this topic.
- These 4 doctests remain disabled, but I've decided not to mock them up either. Why?
  - They are basically wrappers, thin helpers to deal with other module's data. So, first of all, mocking them renders their doctests almost useless (that minimal the functionality is).
  - Additionally, after reviewing `librbd` code I saw it provides `rbd_features_to_string()` and `rbd_features_from_string()`, but unfortunately those aren't exposed as Python bindings. In this case, I think it's better to request those functions to be available instead of re-implementing them on our side. I've opened a [tracker issue to RBD component for this](https://tracker.ceph.com/issues/40833).

epuertat commented:
I summon you: @tchaikov @sebastian-philipp @rjfd @wjwithagen !

I'm currently blocked with this PR due to `install-deps.sh` changes. If I don't modify it, this PR breaks (as `install-deps.sh` does its tricky manual dependency installation instead of running a `tox --no-test` and then `pip --find-links...`). My first approach was to remove the Python caching code, and it worked but with some downsides.

So, in order to deal with `install-deps.sh` I'd need to understand its constraints. If I'm not wrong this is the situation:
- It already installs Python packages with the distro package manager (rpm, deb, pkg, etc.).
  - However, after that, it uses pip for installing Pypi packages that might differ from distro's, and besides virtualenvs would block Python code from using those anyway (this isolation could be disabled with `--system-site-packages`).
- It aims at allowing offline builds without any internet connection:
  - Although Dashboard's NPM installation seems to break that assumption.
  - Based on the Jenkins console logs, it seems that currently some packages are downloaded from the internet.
- It installs both Python2 and 3 dependencies, no matter which Python version is to be tested.

So, if I remove the current pre-population of pip cache, everything works the same and:
- We only install python dependencies when they're required (till now they all were installed upfront no matter you only wanted to build a single component on Py2).
- But we DO need internet connection for the build, and they're not reused anymore between tox envs (which would imply extra bandwidth if we're not caching Pypi at Ceph CI environment).
  - Is this a must?

I'd really appreciate your feedback. Thanks!

sebastian-philipp commented:
> I summon you: @tchaikov @sebastian-philipp @rjfd @wjwithagen !
> 
> I'm currently blocked with this PR due to `install-deps.sh` changes. If I don't modify it, this PR breaks (as `install-deps.sh` does its tricky manual dependency installation instead of running a `tox --no-test` and then `pip --find-links...`). My first approach was to remove the Python caching code, and it worked but with some downsides.

> I'd really appreciate your feedback. Thanks!

Sorry, slightly off topic. It really strikes me that we're creating so many virtual environments. If we could reduce the number of tox environments to one or two, we might get away with less caching. 

There is no need to have so many virtual environments in the first places, as they have to be installable on a single system anyway.  

Some benefits one having only one environment would be:

1. initializing the  tests would be faster
2. less cmake magic
3. easier to add new stuff to cmake
4. less pressure to do extensive caching
5. would assert that we can install everything into one system

sebastian-philipp commented:
> > > Mmm, I'm afraid that is a misconception coming from the fact that we have everything versioned into a single repo, which does not enforce a clear separation of concerns, and also a "make" that makes _literally everything_ (including launching a local cluster).
> > 
> > 
> > Of course, everything else would be wrong. `make` is supposed to make everything.
> 
> `make` is supposed to build code (which is in itself a primary test), ensure a basic degree of sanity testing (unit, component, some integration testing perhaps), and that's all.

Well, that's what it is doing right now?

> 
> > So, running this doc test is not _slow_. Actually it's super fast, if you have compiled ceph libs.
> 
> The trick lies in the "if you have compiled..." clause, which by default equals to `False` (and according to the "build-everything-on-every-change" philosophy, it should be `False` always). So the difference by commenting these 4 doctests (which are trivial and barely test anything in dashboard code), is ~20 minutes build vs. no build at all.

Hmm, `make check` is executed after the libraries are compiled, so as long as the tox run is also integrated in `make check`, there is no minutes overhead.

> 
> I wouldn't comment out these doctests if we took librbd python bindings from Pypi (no trace of it though) and use it directly (though I still think they fall under the definition of integration tests).

In openATTIC we had the joy of being in a separate repository. Honestly, I don't want to go back to that.

epuertat commented:
> There is no need to have so many virtual environments in the first places, as they have to be installable on a single system anyway.
> 
> Some benefits one having only one environment would be:
> 
> 1. initializing the  tests would be faster
> 2. less cmake magic
> 3. easier to add new stuff to cmake
> 4. less pressure to do extensive caching
> 5. would assert that we can install everything into one system

Interesting proposal. I guess the point of virtualenv is isolation, not only from the base system but from other python packages. If we want to just have a common virtualenv for all Python packages, why not simply stop using any virtual environment, as `install-deps.sh` is installing all Python dependencies in the system? I still think this discussion has a lot to do with the downsides of monorepos.

tchaikov commented:
> But we DO need internet connection for the build, and they're not reused anymore between tox envs (which would imply extra bandwidth if we're not caching Pypi at Ceph CI environment).

i am for i think pip has builtin cache machinery. we can use `$XDG_CACHE_HOME` to avoid touching `~/.cache/pip` as ideally, the test should have no side effects.

so i am for "removing the current pre-population of pip cache". as long as we don't write to `$HOME`.

> I guess the point of virtualenv is isolation

also, it allow us to test with the python versions that we want to support even if those python versions are not the "supported" version by the distro on which we run "make check". for instance, we might want to test python3.7, but it is not the supported python in EPEL7, which just moved to python3.6. not to mention that the python modules listed as dependencies in `*requirements*.txt` .

epuertat commented:
> > But we DO need internet connection for the build, and they're not reused anymore between tox envs (which would imply extra bandwidth if we're not caching Pypi at Ceph CI environment).
> 
> i am for i think pip has builtin cache machinery. we can use `$XDG_CACHE_HOME` to avoid touching `~/.cache/pip` as ideally, the test should have no side effects.
> 
> so i am for "removing the current pre-population of pip cache". as long as we don't write to `$HOME`.
> 
> > I guess the point of virtualenv is isolation
> 
> also, it allow us to test with the python versions that we want to support even if those python versions are not the "supported" version by the distro on which we run "make check". for instance, we might want to test python3.7, but it is not the supported python in EPEL7, which just moved to python3.6. not to mention the python modules listed as dependencies in `*requirements*.txt` .

Thanks for the feedback, @tchaikov! I found where the issue came from: `src/tools/setup-virtualenv.sh`.

I think we could easily replace `$XDG_CACHE_HOME` with whatever dir we want to keep caching into.

BTW, I still see in Jenkins `dashboard/orchestrator/insights` Python tests fetching packages from `pypi` repo. We could speed up Jenkins builds by having a shared cache, instead of a per-tox one. Another approach that CircleCI uses is providing key-val blob storge for caching `.tox` dirs and speeding up builds (for example: key=module-vX-<md5sum requirements.txt>-<md5sum constraints.txt>).

epuertat commented:
jenkins test make check arm64

epuertat commented on src/pybind/mgr/dashboard/tox.ini:
```diff
@@ -1,30 +1,56 @@
 skipsdist = true
-toxworkdir = {env:CEPH_BUILD_DIR}/dashboard
-minversion = 2.8.1
+kip_missing_interpreters = true
```
Thanks!

epuertat commented:
jenkins test make check

epuertat commented:
jenkins test make check arm64

epuertat commented:
jenkins test make check arm64

epuertat commented on src/pybind/mgr/dashboard/tests/__init__.py:
```diff
@@ -201,13 +201,13 @@ def run(self):
-            return
+            self.status = 500
+        self.body = json.dumps(thread.res_task['exception'])
+        return
```
Thanks for the catch!

epuertat commented on src/pybind/mgr/dashboard/tox.ini:
```diff
@@ -1,30 +1,66 @@
     CFLAGS = -DXMLSEC_NO_SIZE_T
     UNITTEST = true
     WEBTEST_INTERACTIVE = false
     LD_LIBRARY_PATH = {toxinidir}/../../../../build/lib
```
Mmm, you're right. I'll remove as much as I can from there ;-)

epuertat commented:
> Hi, I still experience some failing tests running the suite [locally](https://paste.opensuse.org/view//50979867) or remotely with [travis-ci](https://pastebin.com/YBgTX4S5).
> 
> Nevertheless great to be finally able to run the unit tests remotely .
> I also like the refactoring you did along the way

It's funny. The error itself makes sense, but cannot understand why those envs report it, while mine (being exactly the same) doesn't:

> platform linux -- Python 3.6.8, pytest-3.10.1, py-1.8.0, pluggy-0.12.0
> rootdir: /ceph/src/pybind/mgr/dashboard, inifile: tox.ini
> plugins: cov-2.7.1, instafail-0.4.1, timeout-1.3.3, spec-1.1.0, flake8-1.0.4
> 
> asn1crypto==0.24.0,atomicwrites==1.3.0,attrs==19.1.0,backports.functools-lru-cache==1.5,bcrypt==3.1.7,certifi==2019.6.16,cffi==1.12.3,chardet==3.0.4,cheroot==6.5.5,CherryPy==18.1.2,coverage==4.5.4,cryptography==2.7,defusedxml==0.5.0,entrypoints==0.3,flake8==3.7.8,idna==2.7,importlib-metadata==0.19,isodate==0.6.0,jaraco.functools==2.0,lxml==4.4.0,mccabe==0.6.1,mock==3.0.5,more-itertools==4.1.0,pkgconfig==1.5.1,pluggy==0.12.0,portend==2.5,py==1.8.0,pycodestyle==2.5.0,pycparser==2.19,pyflakes==2.1.1,PyJWT==1.6.4,pyOpenSSL==17.5.0,pytest==3.10.1,pytest-cov==2.7.1,pytest-faulthandler==2.0.1,pytest-flake8==1.0.4,pytest-instafail==0.4.1,pytest-spec==1.1.0,pytest-timeout==1.3.3,python3-saml==1.4.1,pytz==2019.2,repoze.lru==0.7,requests==2.20.0,Routes==2.4.1,six==1.11.0,tempora==1.14.1,urllib3==1.24.3,xmlsec==1.3.3,zc.lockfile==2.0,zipp==0.5.2

I'll fix the Flake8 issues.

epuertat commented:
Jenkins retest please

epuertat commented:
Jenkins test make check

epuertat commented:
jenkins test dashboard

epuertat commented:
jenkins test dashboard

alfonsomthd commented on src/pybind/mgr/dashboard/tox.ini:
```diff
@@ -1,30 +1,69 @@
-    lint: pylint --rcfile=.pylintrc --jobs=5 . module.py tools.py controllers tests services exceptions.py grafana.py ci/check_grafana_uids.py
-    lint: pycodestyle --max-line-length=100 --exclude=.tox,venv,frontend,.vscode --ignore=E402,E121,E123,E126,E226,E24,E704,W503,E741 .
-    check: python ci/check_grafana_uids.py frontend/src/app ../../../../monitoring/grafana/dashboards
-    run: {posargs}
```
I would like to keep ```run: {posargs}``` as it can be useful for running arbitrary commands.

epuertat replied:
Done, thanks helping fix the python3 mock errors too!

epuertat commented:
jenkins test docs

sebastian-philipp commented on src/pybind/mgr/dashboard/tox.ini:
```diff
@@ -1,30 +1,74 @@
 [testenv]
-setenv=
+deps =
+    -rrequirements.txt
```
Just note, we removed those from `deps =` as a workaround for https://github.com/tox-dev/tox/issues/149 . This change will re-introduce this bug again?

(Well, I'm :+1: for changing this back)

epuertat replied:
Wan't aware of that... As I understand, CI builds won't suffer from this as they will create virtualenvs from scratch. For non-CI builds, we still have the `--recreate` flag to force virtualenv creation.

sebastian-philipp replied:
Hm, forcefully recreating the tox envs make things slower for little gain.

epuertat replied:
Well, venvs would only need to be recreated in case of changing requirements, which is something happening quite sporadically. Besides, the previous approach wouldn't solve all issues either (e.g.: removed packages would remain there). Rather than going back to that workaround, why not exploring alternatives (pipfile) or tox plugins handling this situation?

epuertat commented on src/pybind/mgr/dashboard/HACKING.rst:
```diff
@@ -779,17 +779,17 @@ endpoint:
     @Endpoint(path="/", query_params=['opt1'])
     def index(self, key, opt1, opt2=None):
-      # ...
+      """..."""
```
Interesting. Thanks for the tip! This still has to work with py27 though.

epuertat commented on src/pybind/mgr/dashboard/plugins/feature_toggles.py:
```diff
@@ -132,8 +134,9 @@ def get_controllers(self):
         class FeatureTogglesEndpoint(RESTController):
 
-            def list(_):
+            def list(_):  # pylint: disable=no-self-argument
```
That `self` there would shadow the outer `self`, which is used inside the controller. I'm not very happy with that trick, but it was the simplest way to give plugins API interfacing without re-implementing the `RESTController` magic. An alternative would've been to move `RESTController` pseudo-magic methods to the `Plugin` scope interface, but I wanted to keep the hook-based approach followed by `pluggy`.

sebastian-philipp replied:
ah, thanks! Bummer

epuertat commented on src/pybind/mgr/dashboard/services/exception.py:
```diff
@@ -50,7 +50,7 @@ def wrapper(*args, **kwargs):
 
     # pylint: disable=function-redefined
-    def contextmanager(func):
+    def contextmanager(func):  # noqa: F811
```
line 52 works for pylint, but pep8/pycodestyle/flake8 use the `noqa` token. I think `noqa` only accept codes, not names there.

sebastian-philipp replied:
bummer

epuertat commented on src/pybind/mgr/dashboard/tests/test_api_auditing.py:
```diff
@@ -4,7 +4,10 @@
 import cherrypy
-import mock
+try:
+    import mock
```
In Python 2 `mock` is a standalone package, in Python 3 is under `unittest` tree.

sebastian-philipp replied:
Hm, https://stackoverflow.com/questions/28215214/how-to-add-custom-renames-in-six would be an idea, but I doubt it's worth it.

epuertat replied:
Yeah. I used `six.moves` for other modules here, but if you google for examples, it seems that this is the most common way `mock` module is handled between 2-3 versions.

epuertat commented on src/pybind/mgr/dashboard/conftest.py:
```diff
@@ -0,0 +1,26 @@
+import sys
```
I left that file there because that's the primary location `pytest` looks for, and as this (does/could) also affect to Doctests (we might `mock` the RBD ones there) or other non-`test/` files (e.g.: `ci/`). It's basically initialization/mocking logic common to all testing stuff. Nevertheless I have no problem if someone wants to move this into `test/`.

epuertat commented on src/pybind/mgr/dashboard/.pylintrc:
```diff
@@ -393,7 +393,7 @@ ignored-classes=optparse.Values,thread._local,_thread._local
 # and thus existing member attributes cannot be deduced by static analysis. It
 # supports qualified module names, as well as Unix pattern matching.
-ignored-modules=cherrypy,distutils
+ignored-modules=cherrypy,distutils,rados,rbd,cephfs
```
Adding comments doesn't make any harm, but I'm not sure if that's the reason I'd add. Rather than a matter of build order, I'd say it's a matter of relevance (should we necessarily build 3rd parties to quality-check dashboard code?).
